internal/app/machined/main.go

```diff
@@ -80,217 +79,90 @@ func sync() {
-			return nil
-		}
+func reboot(err error) {
+	log.Print(err)
 
-		return err
+	for i := 10; i >= 0; i-- {
+		log.Printf("rebooting in %d seconds\n", i)
+		time.Sleep(1 * time.Second)
 	}
 
-	// nolint: errcheck
-	defer f.Close()
-
-	adv, err := syslinux.NewADV(f)
-	if err != nil {
-		return err
-	}
-
-	label, ok := adv.ReadTag(syslinux.AdvUpgrade)
-	if !ok {
-		return nil
-	}
-
-	if label == "" {
-		adv.DeleteTag(syslinux.AdvUpgrade)
-
-		if _, err = f.Write(adv); err != nil {
-			return err
-		}
-
-		return nil
-	}
-
-	log.Printf("reverting default boot to %q", label)
-
-	var b []byte
-
-	if b, err = ioutil.ReadFile(syslinux.SyslinuxConfig); err != nil {
-		return err
-	}
-
-	re := regexp.MustCompile(`^DEFAULT\s(.*)`)
-	matches := re.FindSubmatch(b)
-
-	if len(matches) != 2 {
-		return fmt.Errorf("expected 2 matches, got %d", len(matches))
-	}
-
-	b = re.ReplaceAll(b, []byte(fmt.Sprintf("DEFAULT %s", label)))
-
-	if err = ioutil.WriteFile(syslinux.SyslinuxConfig, b, 0600); err != nil {
-		return err
-	}
-
-	adv.DeleteTag(syslinux.AdvUpgrade)
-
-	if _, err = f.Write(adv); err != nil {
-		return err
-	}
+	sync()
 
-	return nil
+	// nolint: errcheck
+	unix.Reboot(unix.LINUX_REBOOT_CMD_RESTART)
 }
 
-// nolint: gocyclo
 func main() {
-	var err error
-
-	// This is main entrypoint into machined execution, control is passed here
-	// from init after switch root.
-	//
-	// When machined terminates either on normal shutdown (reboot, poweroff), or
-	// due to panic, control goes through recovery() and reboot() functions
-	// below, which finalize node state - sync buffers, initiate poweroff or
-	// reboot. Also on shutdown, other deferred function are called, for example
-	// services are gracefully shutdown.
-
-	// On any return from init.main(), initiate host reboot or shutdown handle
-	// any panics in the main goroutine, and proceed to reboot() above
-	defer recovery()
-
-	// Subscribe to events.
-	init := EventBusObserver{&event.Embeddable{}}
-	defer close(init.Channel())
-
-	event.Bus().Register(init)
-
-	defer event.Bus().Unregister(init)
-
-	// Initialize process reaper.
-	reaper.Run()
-	defer reaper.Shutdown()
-
-	// Ensure rng is seeded.
-	if err = startup.RandSeed(); err != nil {
-		panic(err)
-	}
-
-	// Set the PATH env var.
-	if err = os.Setenv("PATH", constants.PATH); err != nil {
-		panic(errors.New("error setting PATH"))
+	p, err := platform.CurrentPlatform()
+	if err != nil {
+		reboot(err)
 	}
 
-	immediateReboot := false
+	r := runtime.NewRuntime(p, nil, runtime.Initialize)
 
-	// Boot the machine.
-	seq := sequencer.New(sequencer.V1Alpha1)
+	s := &v1alpha1runtime.Sequencer{}
 
-	if err := seq.Boot(); err != nil {
-		if errors.Is(err, runtime.ErrReboot) {
-			immediateReboot = true
-		} else {
-			log.Println(err)
-			panic(fmt.Errorf("boot failed: %w", err))
-		}
+	controller := &runtime.Controller{
+		Runtime:   r,
+		Sequencer: s,
 	}
 
-	// Wait for an event.
+	if p.Mode() != runtime.Container {
```

Comment:
probably we could have `p.ShutdownCh() <-chan struct{}` (for each `Platform`) so we hide this complexity behind the `Platform` interface